src/MongoDB.Analyzer/Core/Poco/PocoExpressionProcessor.cs

```diff
@@ -0,0 +1,87 @@
+
+        var typesProcessor = context.TypesProcessor;
+
+        foreach (var classNode in classNodes)
+        {
+            if (analysisContexts.Count == context.Settings.PocoLimit)
+            {
+                break;
+            }
+
+            try
+            {
+                var classSymbol = semanticModel.GetDeclaredSymbol(classNode);
+                if (PreanalyzeClassDeclaration(context, classSymbol))
+                {
+                    var generatedClassName = typesProcessor.ProcessTypeSymbol(classSymbol);
+                    var generatedClassNode = (ClassDeclarationSyntax)(typesProcessor.GetTypeSymbolToMemberDeclarationMapping(classSymbol));
+                    var expresionContext = new ExpressionAnalysisContext(new ExpressionAnalysisNode(classNode, null, generatedClassNode, null, classNode.GetLocation()));
+                    analysisContexts.Add(expresionContext);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"Failed analyzing {classNode.NormalizeWhitespace()} with {ex.Message}");
+            }
+        }
+
+        var pocoAnalysis = new ExpressionsAnalysis()
+        {
+            AnalysisNodeContexts = analysisContexts.ToArray(),
+            InvalidExpressionNodes = invalidExpressionNodes.ToArray(),
+            TypesDeclarations = typesProcessor.TypesDeclarations
+        };
+
+        context.Logger.Log($"JSON: Found {pocoAnalysis.AnalysisNodeContexts.Length} expressions.");
+        return pocoAnalysis;
+    }
+
+    private static bool ContainsBsonAttributes(MongoAnalysisContext context, INamedTypeSymbol classSymbol) =>
+        classSymbol?.GetAttributes().Where(attribute => IsValidBsonAttribute(context, attribute)).AnySafe() ?? false;
+
+    private static bool ContainsPropertiesWithBsonAttributes(MongoAnalysisContext context, INamedTypeSymbol classSymbol) =>
+        classSymbol?.GetMembers().OfType<IPropertySymbol>().SelectMany(property => property.GetAttributes().Where(attribute => IsValidBsonAttribute(context, attribute))).AnySafe() ?? false;
+
+    private static bool ContainsFieldsWithBsonAttributes(MongoAnalysisContext context, INamedTypeSymbol classSymbol) =>
+        classSymbol?.GetMembers().OfType<IFieldSymbol>().SelectMany(field => field.GetAttributes().Where(attribute => IsValidBsonAttribute(context, attribute))).AnySafe() ?? false;
+
+    private static bool IsValidBsonAttribute(MongoAnalysisContext context, AttributeData attribute) =>
+        attribute.AttributeClass.IsSupportedBsonAttribute() || context.TypesProcessor.GetTypeSymbolToMemberDeclarationMapping(attribute.AttributeClass) != null;
+
+    private static bool PreanalyzeClassDeclaration(MongoAnalysisContext context, INamedTypeSymbol classSymbol) =>
+        context.Settings.JsonAnalyzerVerbosity == JsonAnalyzerVerbosity.All || ContainsBsonAttributes(context, classSymbol) ||
```

Comment:
minor: place each condition in its own line. Early exit if `classSymbol == null`, no need to check for `classSymbol?`